### ISSUE TITLE
from_java conversion fails for Strings

### ISSUE THREAD

**Issue by hinerm** [imagej/pyimagej#114]
The following python script:
```
import imagej
ij = imagej.init()
from jpype.types import *
myStr = JString("foo")
py_str = str(myStr) # Works as expected
ij_str = ij.py.from_java(myStr)  # Conversion fails with a Java exception
```

produces this exception:
```
Traceback (most recent call last):
  File "AbstractConvertService.java", line 125, in org.scijava.convert.AbstractConvertService.convert
  File "AbstractDelegateConverter.java", line 53, in org.scijava.convert.AbstractDelegateConverter.convert
  File "AbstractConvertService.java", line 125, in org.scijava.convert.AbstractConvertService.convert
  File "FileToDatasetConverter.java", line 66, in io.scif.convert.FileToDatasetConverter.convert
  File "DefaultDatasetIOService.java", line 138, in io.scif.services.DefaultDatasetIOService.open
  File "DefaultDatasetIOService.java", line 133, in io.scif.services.DefaultDatasetIOService.open
  File "DefaultDatasetIOService.java", line 152, in io.scif.services.DefaultDatasetIOService.open
  File "ImgOpener.java", line 242, in io.scif.img.ImgOpener.openImgs
  File "ImgOpener.java", line 483, in io.scif.img.ImgOpener.createReader
  File "DefaultInitializeService.java", line 87, in io.scif.services.DefaultInitializeService.initializeReader
Exception: Java Exception
```

`JInt` conversion appears to work, as does dataset-related conversion.

Current workaround is to just use the python `str()` method.

**Comment by ctrueden** [comment]
Thanks @hinerm. I'm surprised this wasn't caught in scyjava by the unit tests. We'll need to beef them up accordingly, to cover this case and any similar cases.

**Comment by imagejan** [comment]
> ```
> File "FileToDatasetConverter.java", line 66, in io.scif.convert.FileToDatasetConverter.convert
>  File "DefaultDatasetIOService.java", line 138, in io.scif.services.DefaultDatasetIOService.open
>  File "DefaultDatasetIOService.java", line 133, in io.scif.services.DefaultDatasetIOService.open
>  File "DefaultDatasetIOService.java", line 152, in io.scif.services.DefaultDatasetIOService.open
>  File "ImgOpener.java", line 242, in io.scif.img.ImgOpener.openImgs
>  File "ImgOpener.java", line 483, in io.scif.img.ImgOpener.createReader
>  File "DefaultInitializeService.java", line 87, in io.scif.services.DefaultInitializeService.initializeReader
>Exception: Java Exception
>``` 

This shouldn't happen with `scifio` versions `>=0.41.1` (containing https://github.com/scifio/scifio/commit/30e1749d5667fa8e8e9fff22ec56871648b78dd2 that fixes `FileToDatasetConverter`).

It likely doesn't fix he underlying issue with `DefaultInitializeService` though, which reminds me of https://github.com/scifio/scifio/issues/442.

**Comment by elevans** [comment]
@hinerm I just ran this on my system and my stack trace is a little more verbose than yours (no idea why). I get the same result with the following addition:

```python
The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/edward/Documents/repos/loci/pyimagej/imagej/__init__.py", line 552, in from_java
    raise exc
  File "/home/edward/Documents/repos/loci/pyimagej/imagej/__init__.py", line 545, in from_java
    data = self._ij.convert().convert(data, Dataset)
java.lang.NullPointerException: java.lang.NullPointerException
```

I looked at `scyjava` as it seems we may have just missed casting a string with `str()` but we do in fact do that on line 474-475 in the `scyjava/__init__.py`.

I don't think the string is even getting to scyjava.

**Comment by hinerm** [comment]
Looking at [from_java](https://github.com/imagej/pyimagej/blob/e8fef0b694f0295d85a6f3a57222cdeb4963a642/imagej/__init__.py#L532-L550), it looks like there is an opportunity for the `ConvertService` to be overly aggressive (the `CovnertService` is checked first and has to pass on the input, at which point it passes to scyjava's [to_python](https://github.com/scijava/scyjava/blob/89e1aa9142b71abe2a33aab170894555127ae1b7/scyjava/__init__.py#L416-L499)).

In general we want to use the `to_python` when converting primitives and `ConvertService` for `RandomAccessibleInterval` and `Dataset`. But if we have a `Converter` that converts from primitive type `X` to `RAI` or `Dataset`, the `ConvertService` will always win out.

I assume what's happening here is there is a `Converter` that supports `String` to `RAI` for strings of a certain format, so it says "yes I can convert this", but then it discovers the `String` is not actually formatted for conversion to an `RAI`, producing a `null` `RAI`. Thus the NPE.

We could potentially say try `to_python` first, except that it may also be too aggressive - for example it claims `java.lang.Iterable`, which I assume would convert `net.imagej.Dataset`.

One option is to decouple `to_python` and `from_java`. In this case `to_python` could potentially be removed, relying instead on the native Jpype conversion.

**Comment by ctrueden** [comment]
> I assume what's happening here is there is a Converter that supports String to RAI for strings of a certain format, so it says "yes I can convert this", but then it discovers the String is not actually formatted for conversion to an RAI, producing a null RAI. Thus the NPE.

Could we try the `ConvertService` first, and if it gives `null`, then fall back to scyjava's conversion?

**Comment by imagejan** [comment]
@ctrueden wrote:

> Could we try the `ConvertService` first, and if it gives `null`, then fall back to scyjava's conversion?

If the `ConvertService` returns `null`, its `supports()` method should have returned `false` with the given object, right?
I agree with @hinerm here:

> I assume what's happening here is there is a `Converter` that supports `String` to `RAI` for strings of a certain format, so it says "yes I can convert this"

In this case it is (was) an overly eager `{String|File}ToDatasetConverter` in SCIFIO that claimed *any* string, instead of just those representing an actually existing `File` that can be opened (fixed in latest SCIFIO, but not yet managed by `pom-scijava`).

Nevertheless, I believe it would be good to have some special handling for string conversion to python to prefer `str()` over *any* possible converter to other types, unless explicitly requested, no?

**Comment by hinerm** [comment]
>Nevertheless, I believe it would be good to have some special handling for string conversion to python to prefer str() over any possible converter to other types, unless explicitly requested, no?

I think this could have unintended consequences.. I mean I *think* if given a string that represents a path on disk, opening that dataset is reasonable?

In any case, closing for now since it the issue is resolved as of pyimagej `1.0.2`.. if we identify concrete issues with string conversion we can open a new one!
